internal/pkg/instance/logger.go

```diff
@@ -114,39 +111,68 @@ func (l *Logger) scanOutput(data []byte, atEOF bool) (advance int, token []byte,
 // NewWriter create a new pipe pair for corresponding stream.
 func (l *Logger) NewWriter(stream string, dropCRNL bool) *io.PipeWriter {
 	reader, writer := io.Pipe()
-	go l.scan(stream, reader, writer, dropCRNL)
+	closer := l.scan(stream, reader, writer, dropCRNL)
+	l.Lock()
+	l.closers = append(l.closers, closer)
+	l.Unlock()
 	return writer
 }
 
-func (l *Logger) scan(stream string, pr *io.PipeReader, pw *io.PipeWriter, dropCRNL bool) {
+func (l *Logger) scan(stream string, pr *io.PipeReader, pw *io.PipeWriter, dropCRNL bool) closer {
 	r := strings.NewReplacer("\r", "\\r", "\n", "\\n")
 	scanner := bufio.NewScanner(pr)
 	if !dropCRNL {
 		scanner.Split(l.scanOutput)
 	}
 
-	for scanner.Scan() {
-		l.fileMutex.Lock()
-		if !dropCRNL {
-			fmt.Fprint(l.file, l.formatter(stream, r.Replace(scanner.Text())))
-		} else {
-			fmt.Fprint(l.file, l.formatter(stream, scanner.Text()))
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+
+	go func() {
+		for scanner.Scan() {
+			l.Lock()
+			if !dropCRNL {
+				fmt.Fprint(l.file, l.formatter(stream, r.Replace(scanner.Text())))
+			} else {
+				fmt.Fprint(l.file, l.formatter(stream, scanner.Text()))
+			}
+			l.Unlock()
 		}
-		l.fileMutex.Unlock()
+		pr.Close()
+		wg.Done()
+	}()
+
+	// closer function
+	return func() {
+		pw.Close()
+		wg.Wait()
+	}
+}
+
+// Close closes all pipe pairs created with NewWriter and also closes
+// log file descriptor.
+func (l *Logger) Close() {
+	// closer will terminate scan goroutines spawned with NewWriter
+	// by closing pipe write end
+	for _, closer := range l.closers {
+		closer()
 	}
 
-	pr.Close()
-	pw.Close()
+	l.Lock()
+	l.closers = nil
+	l.file.Sync()
+	l.file.Close()
+	l.Unlock()
 }
 
 // ReOpenFile closes and re-open log file (eg: log rotation).
 func (l *Logger) ReOpenFile() {
-	l.fileMutex.Lock()
-	defer l.fileMutex.Unlock()
-
-	path := l.file.Name()
-
+	l.Lock()
+	l.file.Sync()
 	l.file.Close()
```

Comment:
this bit here invalidates the file handle and releases the lock. Since l.Unlock() is a goroutine yield point, it's entirely possible that the scan gorotines waiting on the lock will get unblocked and proceed trying to write to the now-closed file handle.

This means that openFile has to be changed to that it must be called with the lock held.